airflow/models/taskmap.py

```diff
@@ -0,0 +1,71 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Table to store "mapped" task instances (AIP-42)."""
+
+import enum
+
+from sqlalchemy import Column, ForeignKeyConstraint, Integer, String
+
+from airflow.models.base import COLLATION_ARGS, ID_LEN, Base
+from airflow.utils.sqlalchemy import ExtendedJSON
+
+
+class TaskMapVariant(enum.Enum):
+    """Task map variant.
+
+    Possible values are **dict** (for a key-value mapping) and **list** (for an
+    ordered value sequence).
+    """
+
+    DICT = "dict"
+    LIST = "list"
+
+
+class TaskMap(Base):
+    """Model to track dynamic task-mapping information.
+
+    This is currently only populated by an upstream TaskInstance pushing an
+    XCom that's pulled by a downstream for mapping purposes.
+    """
+
+    __tablename__ = "task_map"
+
+    # Link to upstream TaskInstance creating this dynamic mapping information.
+    dag_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True)
+    task_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True)
+    run_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True)
+    map_id = Column(Integer, primary_key=True)
```

Comment:
Maybe we should call this `map_index` -- The `id` suffix to me implies its a FK to something, but this isn't an ID itself, but an index in to some other list.

Reply:
Changed